wire-ios-request-strategy/Sources/Request Strategies/Conversation/MLSConversationParticipantsService.swift

```diff
@@ -0,0 +1,141 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireDataModel
+
+// sourcery: AutoMockable
+protocol MLSConversationParticipantsServiceInterface {
+
+    func addParticipants(
+        _ users: [ZMUser],
+        to conversation: ZMConversation
+    ) async throws
+
+    func removeParticipant(
+        _ user: ZMUser,
+        from conversation: ZMConversation
+    ) async throws
+
+}
+
+enum MLSConversationParticipantsError: Error, Equatable {
+    case ignoredUsers(users: Set<ZMUser>)
+    case invalidOperation
+}
+
+struct MLSConversationParticipantsService: MLSConversationParticipantsServiceInterface {
+
+    // MARK: - Properties
+
+    private let context: NSManagedObjectContext
+    private let clientIDsProvider: MLSClientIDsProviding
+    private let mlsService: MLSServiceInterface
+
+    // MARK: - Life cycle
+
+    init?(context: NSManagedObjectContext) {
+
+        guard let syncContext = context.performAndWait({ context.zm_sync }) else {
+            return nil
+        }
+
+        guard let mlsService = syncContext.performAndWait({ syncContext.mlsService }) else {
+            return nil
+        }
+
+        self.init(
+            context: context,
+            mlsService: mlsService,
+            clientIDsProvider: MLSClientIDsProvider()
+        )
+    }
+
+    init(
+        context: NSManagedObjectContext,
+        mlsService: MLSServiceInterface,
+        clientIDsProvider: MLSClientIDsProviding
+    ) {
+        self.context = context
+        self.mlsService = mlsService
+        self.clientIDsProvider = clientIDsProvider
+    }
+
+    // MARK: - Interface
+
+    func addParticipants(
+        _ users: [ZMUser],
+        to conversation: ZMConversation
+    ) async throws {
+
+        let (qualifiedID, groupID) = await context.perform {
+            (conversation.qualifiedID, conversation.mlsGroupID)
+        }
+
+        Logging.mls.info("adding \(users.count) participants to conversation (\(String(describing: qualifiedID)))")
+
+        guard let groupID else {
+            Logging.mls.warn("failed to add participants to conversation (\(String(describing: qualifiedID))): missing group ID")
+            throw MLSConversationParticipantsError.invalidOperation
+        }
+
+        let mlsUsers = await context.perform { users.compactMap(MLSUser.init(from: )) }
+
+        do {
+            try await mlsService.addMembersToConversation(with: mlsUsers, for: groupID)
+        } catch MLSService.MLSAddMembersError.failedToClaimKeyPackages {
+            // TODO: Refactor key package claim in `MLSService` to throw an error with the users that couldn't be added
```

Comment:
How can we keep track of this TODO?

Reply:
I will do this in another PR for the same ticket as the one for this PR https://wearezeta.atlassian.net/browse/WPB-529

Reply:
@netbe here